### ISSUE TITLE
take(0) doesn't have any effect if the previous operator doesn't complete

### ISSUE THREAD

**Issue by stavalfi** [reactor/reactor-core#1158]
```java
List<Integer> block = Flux/Mono.just(1)
        .flatMap/flatMapMany(number -> Flux/Mono.<Integer>create(sink -> {
              // In my real app - this stream will never complete but may signal error.
        }))
        .take(0)
        .collectList()
        .block();
System.out.println(block); // we never come here
```
I have tests which inside the flatMap I wait for incoming messages and everytime the number inside the `take` is different. In the test where I have `take(0)`, the test fail due to this reason.

### Expected behavior
`[]`
### Actual behavior
nothing is printed
### Steps to reproduce
copy the above code.
### Reactor Core version
`Bismuth-SR7`
### JVM version (e.g. `java -version`)
`1.8`

**Comment by simonbasle** [comment]
`take` works by hooking in `onNext` events, cancelling the upstream when it has seen enough elements. `n = 0` is a corner case that is so far handled by [immediately requesting upon subscription](https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/FluxTake.java#L94), which would trigger the `take` cancellation upon receiving the first value.

However, `never()` like sources like yours, which **emit no signals**, are a corner case. Any source that never completes and emits less elements than the `n` configured for `take` would be problematic, but the `0` case at least can be solved: we could cancel right away on subscription instead of triggering that `request(Long.MAX_VALUE)`.

So TL;DR
 - we can fix the `Flux.never().take(0)` case (cancel instead of request upon subscription)
 - we cannot fix the `Flux.never().take(1)` case, avoiding this is the responsibility of the users

**Comment by akarnokd** [comment]
In RxJava, we cancel in `onSubscribe` if the limit is zero: [FlowableTake.java](https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/operators/flowable/FlowableTake.java#L52).

**Comment by simonbasle** [comment]
@akarnokd yes this approach is the better one, covers the `never()`-like sources nicely 👍

**Comment by stavalfi** [comment]
@simonbasle Why you consider `Flux.never().take(1)` as a reactor-core bug?
